beacon_node/eth2_libp2p/src/peer_manager/mod.rs

```diff
@@ -940,19 +947,31 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
         let connected_peer_count = self.network_globals.connected_peers();
         if connected_peer_count > self.target_peers {
             //remove excess peers with the worst scores, but keep subnet peers
-            for (peer_id, _) in self
+            //must also ensure that the outbound-only peer count does not go below the minimum threshold
+            let mut peers_removed = 0;
+            let mut n_outbound_removed = 0;
+            for (peer_id, info) in self
                 .network_globals
                 .peers
                 .read()
                 .worst_connected_peers()
                 .iter()
                 .filter(|(_, info)| !info.has_future_duty())
-                .take(connected_peer_count - self.target_peers)
-                //we only need to disconnect peers with healthy scores, since the others got already
-                //disconnected in update_peer_scores
-                .filter(|(_, info)| info.score_state() == ScoreState::Healthy)
             {
-                disconnecting_peers.push(**peer_id);
+                if peers_removed == connected_peer_count - self.target_peers {
+                    break;
+                }
+                if info.is_outbound_only() {
+                    if n_outbound_removed < outbound_only_peer_count - min_outbound_only_target {
```

Comment:
Looks to me like this can underflow if 
`connected_peer_count  > target_peers` but `outbound_only_peer_count` > `min_outbound_only_target` 

Probably safer to use safe math every and log the errors. We'd probably want to not do this calculation if this is the case tho, so make it conditional that `outbound_only_peer_count < min_outbound_only_target`

Reply:
I might be interpreting you incorrectly, but if we use that as the conditional, isn't it possible that we could have removed some "bad" outbound-only peers, but do not because the conditional evaluates as false and then we iterate to the next. ie. in the test case I made, if we had another outbound-only peer that also had a score of -1, the two outbound-only peers will have the worst scores, but because ```outbound_only_peer_count > min_outbound_only_target```, we end up removing 2 peers with better scores but are not outbound-only. With the current logic it would instead remove one outbound-only peer and one other peer, so if an outbound-only peer has a bad score we would be able to remove it, while still meeting the quota.

Reply:
Yes. I think you may have missed mine and Diva's point with our comments.
In rust, this statement: `outbound_only_peer_count - min_outbound_only_target ` will panic if `outbound_only_peer_count` is < `min_outbound_only_target` and will crash the lighthouse client. 

I'm suggesting to not do this subtraction in this case to avoid crashing the client. Rust has some SafeMath alternatives such as `outbound_only_peer_count.saturating_sub(min_outbound_only_target)` which won't panic but will saturate to 0.

Reply:
Yup just realized what you guys meant. Thanks for pointing it out! Was caught up in thinking through the logic that I didn't consider that these are unsigned!

Reply:
Would changing the conditional to ```min_outbound_only_target < outbound_only_peer_count - n_outbound_removed``` preserve the logic and handle the overflow issue? ```n_outbound_removed``` is bounded above by ```outbound_only_peer_count``` so then we should be able to avoid the issue without needing to use SafeMath?

Reply:
Regardless I will add some more test cases since yea that would have been a pretty big issue

Reply:
Yep, this seems like a sensible solution :).